src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,395 @@
+                                                  Schedule(actor);
+                                                }));
+}
+
+std::vector<ActorID> GcsActorScheduler::CancelOnNode(const ClientID &node_id) {
+  // Remove all the actors from the map associated with this node, and return them as they
+  // will be reconstructed later.
+  std::vector<ActorID> actor_ids;
+
+  // Remove all actors in phase of leasing.
+  {
+    auto iter = node_to_actors_when_leasing_.find(node_id);
+    if (iter != node_to_actors_when_leasing_.end()) {
+      actor_ids.insert(actor_ids.end(), iter->second.begin(), iter->second.end());
+      node_to_actors_when_leasing_.erase(iter);
+    }
+  }
+
+  // Remove all actors in phase of creating.
+  {
+    auto iter = node_to_workers_when_creating_.find(node_id);
+    if (iter != node_to_workers_when_creating_.end()) {
+      for (auto &entry : iter->second) {
+        actor_ids.emplace_back(entry.second->TakeAssignedActorID());
+      }
+      node_to_workers_when_creating_.erase(iter);
+    }
+  }
+
+  // Remove the related remote lease client from remote_lease_clients_.
+  remote_lease_clients_.erase(node_id);
+  return actor_ids;
+}
+
+ActorID GcsActorScheduler::CancelOnWorker(const ClientID &node_id,
+                                          const WorkerID &worker_id) {
+  // Remove the worker which is creating actor and return the related actor.
+  return RemoveWorkerInPhaseOfCreating(node_id, worker_id);
+}
+
+void GcsActorScheduler::LeaseWorkerFromNode(std::shared_ptr<GcsActor> actor,
+                                            std::shared_ptr<rpc::GcsNodeInfo> node) {
+  RAY_CHECK(actor && node);
+
+  auto node_id = ClientID::FromBinary(node->node_id());
+  RAY_LOG(DEBUG) << "Start leasing worker from node " << node_id << " for actor "
+                 << actor->GetActorID();
+
+  auto retry_leasing_worker = [this, node_id, actor] {
+    // Reset the client so that we can reconnect when retry.
+    remote_lease_clients_.erase(node_id);
+    execute_after(io_context_,
+                  [this, node_id, actor] {
+                    if (auto node = gcs_node_manager_.GetNode(node_id)) {
+                      LeaseWorkerFromNode(actor, node);
+                    }
+                  },
+                  RayConfig::instance().gcs_lease_worker_retry_interval_ms());
+  };
+
+  rpc::Address remote_address;
+  remote_address.set_raylet_id(node->node_id());
+  remote_address.set_ip_address(node->node_manager_address());
+  remote_address.set_port(node->node_manager_port());
+  auto lease_client = GetOrConnectLeaseClient(remote_address);
+  auto status = lease_client->RequestWorkerLease(
+      actor->GetCreationTaskSpecification(),
+      [this, node_id, retry_leasing_worker, actor](
+          const Status &status, const rpc::RequestWorkerLeaseReply &reply) {
+        if (auto node = gcs_node_manager_.GetNode(node_id)) {
+          if (!status.ok()) {
+            retry_leasing_worker();
+            return;
```

Comment:
> Nit, can you use an if-else instead of an early return here?

This is the programming habit that I have cultivated for so many years. I will pay attention to it later. 🙂